### PR TITLE
[RFC] Asynchronous notifications

### PR DESCRIPTION
**Pull request by jenswi-linaro**
See PR https://github.com/OP-TEE/optee_os/pull/4601

### CODE REVIEW

File: include/linux/tee_drv.h [hunk 1]
```diff
@@ -466,6 +466,34 @@ static inline int tee_shm_get_id(struct tee_shm *shm)
  */
 struct tee_shm *tee_shm_get_from_id(struct tee_context *ctx, int id);
 
+
+/**
+ * tee_dev_open_helper() - helper function to make a struct tee_context
+ * @teedev:	Device to open
+ *
+ * Creates the struct tee_context without increasing the refernce counter
```
Comment (jforissier) [inline include/linux/tee_drv.h]: reference

File: include/linux/tee_drv.h [hunk 2]
```diff
@@ -466,6 +466,34 @@ static inline int tee_shm_get_id(struct tee_shm *shm)
...
+ * on @teedev. This is needed for instance when a driver need an internal
+ * struct tee_context to operate on. By skipping the reference counter
+ * the circular dependency is broken.
+ *
+ * Note that this struct tee_context need special care when freeing to in
```
Comment (jforissier) [inline include/linux/tee_drv.h]: s/to //

File: drivers/tee/optee/optee_smc.h [hunk 3]
```diff
@@ -107,6 +107,12 @@ struct optee_smc_call_get_os_revision_result {
 /*
  * Call with struct optee_msg_arg as argument
  *
+ * When calling this function normal world has a few responisbilities:
```
Comment (jforissier) [inline drivers/tee/optee/optee_smc.h]: responsibilities

**Review by jforissier — COMMENTED** [review]
Regarding "tee: separate notification functions":

> The keys are currently only representing secure world threads which
> number usually are never even close to 63 so it should be safe for now.

Arm servers with 128 cores do exist. Although a TEE is not historically a server thing, and although having one secure thread per CPU core is certainly not mandatory, I can already imagine some people wanting to set `CFG_NUM_THREADS=128` ;-)

Cosmetic comments below. Are you planning on posting this RFC to the LKML?

**Comment by jenswi-linaro** [comment]
> Regarding "tee: separate notification functions":
> 
> > The keys are currently only representing secure world threads which
> > number usually are never even close to 63 so it should be safe for now.
> 
> Arm servers with 128 cores do exist. Although a TEE is not historically a server thing, and although having one secure thread per CPU core is certainly not mandatory, I can already imagine some people wanting to set `CFG_NUM_THREADS=128` ;-)


Good point, should we make this 255 instead?

> 
> Cosmetic comments below. Are you planning on posting this RFC to the LKML?

Yes, any day now so the comments are well timed.

**Comment by jenswi-linaro** [comment]
Update. I don't expect any tags in this PR, eventual tags can be saved to when I've posted this on LKML.

**Comment by jforissier** [comment]
> > Regarding "tee: separate notification functions":
> > > The keys are currently only representing secure world threads which
> > > number usually are never even close to 63 so it should be safe for now.
> > 
> > 
> > Arm servers with 128 cores do exist. Although a TEE is not historically a server thing, and although having one secure thread per CPU core is certainly not mandatory, I can already imagine some people wanting to set `CFG_NUM_THREADS=128` ;-)
> 
> Good point, should we make this 255 instead?

Doing so won't use much more memory, right? So it would be preferable I think. I guess you're bound to getting comments whatever value you choose anyways. Couldn't this max key value be queried from the TEE?

**Comment by jenswi-linaro** [comment]
> > > Regarding "tee: separate notification functions":
> > > > The keys are currently only representing secure world threads which
> > > > number usually are never even close to 63 so it should be safe for now.
> > > 
> > > 
> > > Arm servers with 128 cores do exist. Although a TEE is not historically a server thing, and although having one secure thread per CPU core is certainly not mandatory, I can already imagine some people wanting to set `CFG_NUM_THREADS=128` ;-)
> > 
> > 
> > Good point, should we make this 255 instead?
> 
> Doing so won't use much more memory, right? So it would be preferable I think. I guess you're bound to getting comments whatever value you choose anyways. 

It will cost as many bits, nothing more. :-)

> Couldn't this max key value be queried from the TEE?

It can, but not before the counterpart (https://github.com/OP-TEE/optee_os/pull/4601) is available in secure world so this is a just a temporary bridge while upgrading the different software components.

**Pull request closed without merging**